kuma/users/adapters.py

```diff
@@ -207,17 +210,91 @@ def get_connect_redirect_url(self, request, socialaccount):
+        # class is used when calling this method from the one used when calling
+        # is_auto_signup_allowed().
+        user = get_existing_user(sociallogin)
+        if user:
+            # We can re-use an existing user instead of creating a new one.
+            # Let's guarantee this user has an unusable password, just in case
+            # we're recovering an old user that has never had this done before.
+            user.set_unusable_password()
+            # This associates this new social account with the existing user.
+            sociallogin.connect(request, user)
+            # Since the "connect" call above does not add any email addresses
+            # from the social login that are missing from the user's current
+            # associated set, let's add them here.
+            add_user_email(request, user, sociallogin.email_addresses)
+            # Now that we've successfully associated a GitHub/Google social
+            # account with this existing user, let's delete all of the user's
+            # associated Persona social accounts (if any). Users may have
+            # multiple associated Persona social accounts (each identified
+            # by a unique email address).
+            user.socialaccount_set.filter(provider='persona').delete()
+        else:
+            user = super().save_user(request, sociallogin, form)
+
         try:
             del request.session['socialaccount_sociallogin']
         except KeyError:  # pragma: no cover
             pass
+
+        return user
+
+    def is_auto_signup_allowed(self, request, sociallogin):
+        """
+        We allow "auto-signup" (basically skipping the sign-up form page) only
+        if there is an existing user that we can re-use instead of creating
+        a new one. An existing user is found if at least one of the user's
+        verified email addresses matches one of the verified email addresses
+        within the social login object.
+        """
+        return bool(get_existing_user(sociallogin))
+
+
+def get_existing_user(sociallogin):
+    """
+    Attempts to find an existing user that is associated with a verified email
+    address that matches one of the verified email addresses within the
+    "sociallogin" object.
+    """
+    emails = Q()
+    for email_address in sociallogin.email_addresses:
+        if email_address.verified:
+            emails |= Q(emailaddress__email=email_address.email)
+    if emails:
+        # Users can have multiple associated EmailAddress objects, so
+        # let's use "distinct()" to remove any duplicate users.
+        users = list(get_user_model().objects
+                                     .filter(emails,
+                                             emailaddress__verified=True)
+                                     .distinct())
+        # For now, we're only going to return a user if there's only one.
+        if len(users) == 1:
+            return users[0]
```

Comment:
This does make it impossible for users to choose which email address of their social_account they want to use if they already have an account with any of their email addresses, right? So we are sacrificing some freedom for some sanity? Which might be the right trade-off to make, just want to make sure I'm being conscious of it.

If I were to dream of the optimal UX, it'd probably be something like:
- if you only have one verified email and there exists only one account for it, do what we're doing here
- if you have multiple verified emails, pick one
  - if there's only one account account for it, do what we're doing here
  - if there's multiple, pick one

Is that really optimal? I don't know, it clutters login/signup and puts a lot more work in our lap. I think this trade-off might be good enough for now, until people start complaing. Just wanted to think this through for a second 🤔

Reply:
Great points and questions @Gregoor.

> This does make it impossible for users to choose which email address of their social_account they want to use if they already have an account with any of their email addresses, right? So we are sacrificing some freedom for some sanity?

Yes, you're right on. I had originally gone down the path of assuming that even for the case when we find an existing user, we should still go to the signup page so that the user has the chance to change her/his username/email, but that's more challenging to implement (for various reasons) and it wasn't what @atopal wanted. However, even given the approach of this PR, the user always has the option of editing his/her profile to create/select a new primary email address.

Reply:
@Gregoor I wasn't sure what to do for the case when we find multiple user accounts that are associated with one or more verified social-provider emails, so I decided to not pick one at all. The downside, as you see, is that means another, brand-new user account is created for the user. Perhaps that's the wrong decision, and that instead we should allow the user to select one (and perhaps optionally delete the others?) via a separate page? What do you think @atopal?

Reply:
@escattone that's a legacy issue, right? From here on, if you authenticate with emails we already have an account for, we'll add that to the existing account. Is that correct? I'm not too worried about the case with multiple accounts with the same email. We can handle those cases manually at first and then automate if necessary. I just wonder what account we log you in with when you log in (just out of curiosity).

Reply:
@atopal Thanks for taking some time to comment, as I know you're busy with your trip! Yes, it's a legacy issue, and from here on we'll add to an existing account. Sounds good, i think you're right that this is a rare case, if it exists at all, and that we can manually handle it if an issue arises.